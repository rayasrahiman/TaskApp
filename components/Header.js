import React from 'react';
import {
  View,
  Text,
  Image,
  StyleSheet,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from 'react-native-responsive-screen';
import {Colors} from '../constants/colors';

export default function Header({
  dashboard,
  headerRight,
  onPress,
  title,
  ...headerProps
}) {
  return (
    <View
      {...headerProps}
      style={{
        ...styles.container,
        ...(dashboard && {
          marginTop: Dimensions.get('window').height < 800 ? -hp(11) : -hp(5),
        }),
      }}>
      <View style={styles.imgCont}>
        <Image
          source={require('../assets/Logo.png')}
          style={styles.img}
          resizeMode="contain"
        />
      </View>
      {headerRight && (
        <TouchableOpacity onPress={onPress} style={styles.btnCont}>
          <Text style={styles.btnTxt}>{title}</Text>
        </TouchableOpacity>
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    // justifyContent: 'center',
    alignItems: 'center',
    borderBottomColor: Colors.gray,
    borderBottomWidth: wp(0.3),
    width: Dimensions.get('window').width,
    height: hp(15),
  },
  imgCont: {
    marginLeft: wp(40),
  },
  img: {
    width: wp(15),
    height: hp(15),
  },
  btnCont: {
    marginRight: wp(5),
  },
  btnTxt: {
    fontSize: hp(3),
    color: Colors.lightBlue,
  },
});
