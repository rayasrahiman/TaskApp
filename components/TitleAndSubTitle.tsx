import {View, Text, StyleSheet, Platform} from 'react-native';
import React from 'react';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from 'react-native-responsive-screen';

import {Colors} from '../constants/colors';

export type Props = {
  title: string;
  subTitle: string;
  titleContProp: any;
  titleTxtProp: any;
  subTitleContProp: any;
  emailSubTitleContProp: any;
  subTitleTxtProp: any;
  viewPager: boolean;
  email: boolean;
  emailVal: string;
};

const TitleAndSubTitle: React.FC<Props> = ({
  title,
  subTitle,
  titleContProp,
  titleTxtProp,
  subTitleContProp,
  emailSubTitleContProp,
  subTitleTxtProp,
  viewPager,
  email,
  emailVal,
  ...titleAndSubTitleProp
}) => {
  return (
    <View {...titleAndSubTitleProp}>
      <View
        style={{
          ...{
            marginHorizontal: viewPager
              ? 0
              : Platform.OS === 'android'
              ? wp(5)
              : wp(3),
          },
          ...titleContProp,
        }}>
        <Text style={{...styles.titleTxt, ...titleTxtProp}}>{title}</Text>
      </View>
      <View
        style={{
          ...{
            marginHorizontal: viewPager
              ? 0
              : Platform.OS === 'android'
              ? wp(5)
              : wp(3),
          },
          ...subTitleContProp,
        }}>
        <Text
          style={{
            ...styles.subTitleTxt,
            ...{color: Colors.grayishBlack},
            ...subTitleTxtProp,
          }}>
          {subTitle}
        </Text>
      </View>
      {email && (
        <View
          style={{
            ...{
              marginHorizontal: viewPager
                ? 0
                : Platform.OS === 'android'
                ? wp(5)
                : wp(3),
              marginBottom: hp(2),
            },
            ...emailSubTitleContProp,
          }}>
          <Text
            style={{
              ...styles.subTitleTxt,
              ...{color: Colors.black},
              ...subTitleTxtProp,
            }}>
            {emailVal}
          </Text>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  titleTxt: {
    color: Colors.darkBlue,
    fontWeight: 'bold',
    fontSize: Platform.OS === 'android' ? hp(5) : hp(4),
  },
  subTitleTxt: {
    color: Colors.grayishBlack,
    fontWeight: 'bold',
    fontSize: Platform.OS === 'android' ? hp(3) : hp(2),
  },
});

export default TitleAndSubTitle;
