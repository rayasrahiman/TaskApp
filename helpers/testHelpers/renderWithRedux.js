import React from 'react';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {createStore, combineReducers, applyMiddleware} from 'redux';
import createSagaMiddleware from 'redux-saga';

import myFirstReducer from '../../redux/reducer';
import rootSaga from '../../redux/sagas';

export const renderWithRedux = renderedComp => {
  const sagaMiddleware = createSagaMiddleware();
  const rootReducer = combineReducers({myFirstReducer});

  const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
  sagaMiddleware.run(rootSaga);

  return render(<Provider store={store}>{renderedComp}</Provider>);
};
