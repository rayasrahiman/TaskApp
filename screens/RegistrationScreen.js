import React, {useState, useEffect} from 'react';
import {View, Text, StyleSheet, Dimensions} from 'react-native';
import {heightPercentageToDP as hp} from 'react-native-responsive-screen';
import {useSelector, useDispatch} from 'react-redux';

import Button from '../components/Button';
import Header from '../components/Header';
import Input from '../components/Input';
import TitleAndSubTitle from '../components/TitleAndSubTitle';
import I18n from '../languages/i18n';
import {getUserFetch} from '../redux/actionsConstants';

export default function RegistrationScreen({navigation}) {
  const [input, setInput] = useState('');
  const [disable, setDisable] = useState(true);
  const [error, setError] = useState(false);
  const users = useSelector(state => state.myFirstReducer.users);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getUserFetch());
  }, [dispatch]);

  const regex =
    /^(([^<>()[\].,;:\s@"]+(\.[^<>()[\].,;:\s@"]+)*)|(".+"))@(([^<>()[\].,;:\s@"]+\.)+[^<>()[\].,;:\s@"]{2,})$/i;

  const validate = text => {
    setInput(text);
    setDisable(!regex.test(text));
  };

  const redirect = async () => {
    const user = users.find(item => item.email === input);
    if (!disable && user) {
      navigation.navigate('SignIn', {
        email: input,
      });
    } else if (!disable && !user) {
      navigation.navigate('Register', {
        email: input,
      });
    } else {
      setError(true);
    }
  };

  return (
    <View style={styles.mainWrapper}>
      <Header testID="header" />
      <TitleAndSubTitle
        testID="titleSubTitle"
        title={I18n.t('registrationTitle')}
        subTitle={I18n.t('registrationSubTitle')}
        subTitleContProp={styles.subTitleCont}
      />
      <Input
        testID="emailID"
        editable={true}
        label={I18n.t('Email')}
        placeholder={I18n.t('Email')}
        onChangeText={validate}
        error={error && <Text>{I18n.t('EmailError')}</Text>}
        onFocus={() => setError(false)}
      />
      <Button
        testID="register"
        onPress={redirect}
        title={I18n.t('GetStarted')}
        btnWrapper={styles.btnWrapper}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  mainWrapper: {
    flex: 1,
  },
  subTitleCont: {
    marginVertical: hp(2),
  },
  btnWrapper: {
    marginTop: Dimensions.get('window').height < 700 ? hp(30) : hp(30),
  },
});
