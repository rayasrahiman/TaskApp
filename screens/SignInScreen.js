import React, {useState, useEffect} from 'react';
import {
  View,
  Text,
  KeyboardAvoidingView,
  Platform,
  Dimensions,
  ScrollView,
  StyleSheet,
} from 'react-native';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from 'react-native-responsive-screen';
import AsyncStorage from '@react-native-async-storage/async-storage';

import Header from '../components/Header';
import Button from '../components/Button';
import Input from '../components/Input';
import TitleAndSubTitle from '../components/TitleAndSubTitle';
import {Colors} from '../constants/colors';
import I18n from '../languages/i18n';

export default function SignInScreen({navigation, route}) {
  const [error, setError] = useState(false);
  const [loader, setLoader] = useState(false);
  const [inputValue, setInputValue] = useState(false);
  const [pass, setPass] = useState('');
  const {email} = route.params;

  const users = async () => {
    const arr = await AsyncStorage.getItem('users');
    console.log(JSON.parse(arr));
  };

  useEffect(() => {
    users();
    if (email) {
      setInputValue(true);
    }
  }, []);

  const onChange = text => {
    setPass(text);
  };

  const redirect = async () => {
    const arr = await AsyncStorage.getItem('users');
    const array = arr ? JSON.parse(arr) : [];
    const user = array.find(item => item.email === email);
    if (user.password === pass) {
      setLoader(true);
      navigation.navigate('Dashboard');
      setLoader(false);
    } else {
      setError(true);
    }
  };

  return (
    <View style={styles.container}>
      <Header />
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.container}>
        <ScrollView>
          <TitleAndSubTitle
            title={I18n.t('signInTitle')}
            subTitle={I18n.t('signInSubTitle')}
            subTitleContProp={styles.subTitleCont}
          />
          <Input
            editable={false}
            label={I18n.t('Email')}
            placeholder={I18n.t('Email')}
            value={email}
            inputValue={inputValue}
            titleHeaderProp={inputValue && {color: Colors.grayishBlack}}
          />
          <Input
            password
            editable={true}
            label={I18n.t('Password')}
            placeholder={I18n.t('Enterpassword')}
            onChangeText={onChange}
            error={error && <Text>{I18n.t('PasswordError')}</Text>}
            onFocus={() => setError(false)}
            passwordLoader={loader}
            forgotPassword={true}
            forgotPass={() => navigation.navigate('ForgotPassword')}
          />

          <Button
            testID="signIn"
            title={I18n.t('Signin')}
            btnWrapper={{...styles.btnCont, ...styles.btnFirst}}
            onPress={redirect}
          />
          <Button
            title={I18n.t('SwitchAccount')}
            btnWrapper={{...styles.btnCont, ...styles.btnSecond}}
            btnContainerProp={styles.btnSecondCont}
            btnTxtProp={styles.btnSecondTxt}
            onPress={() => {}}
          />
        </ScrollView>
      </KeyboardAvoidingView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  subTitleCont: {
    marginVertical: hp(2),
  },
  btnCont: {
    flex: Platform.OS === 'ios' ? 1 : 0,
    justifyContent: Platform.OS === 'ios' ? 'flex-end' : null,
    alignItems: Platform.OS === 'ios' ? 'flex-end' : null,
  },
  btnFirst: {
    marginTop:
      Platform.OS === 'android' && Dimensions.get('window').height < 700
        ? hp(8.5)
        : hp(12.5),
  },
  btnSecond: {
    marginVertical: hp(3),
    marginHorizontal: wp(5),
  },
  btnSecondCont: {
    backgroundColor: Colors.white,
    borderColor: Colors.lightBlue,
    borderWidth: wp(0.3),
  },
  btnSecondTxt: {
    color: Colors.lightBlue,
  },
});
