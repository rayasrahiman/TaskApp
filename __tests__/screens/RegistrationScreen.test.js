import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react-native';
import {shallow} from 'enzyme';
import {useNavigation} from '@react-navigation/native';

import {renderWithRedux} from '../../helpers/testHelpers/renderWithRedux';
import RegistrationScreen from '../../screens/RegistrationScreen';
import Header from '../../components/Header';
import TitleAndSubTitle from '../../components/TitleAndSubTitle';
import SignInScreen from '../../screens/SignInScreen';

beforeEach(jest.spyOn(console, 'error').mockImplementation(() => undefined));
afterEach(() => jest.clearAllMocks());

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: jest.fn(),
  };
});

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

describe('RegistrationScreen', () => {
  test('should render Registration screen correctly', () => {
    const wrapper = renderWithRedux(<RegistrationScreen />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  test('should render Header', () => {
    const wrapper = renderWithRedux(<RegistrationScreen />);
    wrapper.getByTestId('header');
  });

  test('should render TitleAndSubTitle component', () => {
    const wrapper = renderWithRedux(<RegistrationScreen />);
    wrapper.getByTestId('titleSubTitle');
  });

  test('should render TitleAndSubTitle texts', () => {
    const wrapper = renderWithRedux(<RegistrationScreen />);
    wrapper.getByText('Welcome to the Pentair Home app!');
    wrapper.getByText(
      "Get the most out of your home's water. Enter your email to get started.",
    );
  });

  test('should render label', () => {
    const wrapper = renderWithRedux(<RegistrationScreen />);
    wrapper.getByText('Get Started');
  });

  // test('should show invalid message when invalid email entered', () => {
  //   const {getByTestId, getByText} = render(<RegistrationScreen />);
  //   fireEvent.press(getByTestId('register'));
  //   getByText('Enter a valid email.');
  // });

  test('should navigate to SignIn Screen with email', async () => {
    const mockNavigate = jest.fn();
    useNavigation.mockReturnValueOnce({navigate: mockNavigate});

    const wrapper = renderWithRedux(<RegistrationScreen />);
    // wrapper.find('Button').simulate('click');
    fireEvent.press(wrapper.getByTestId('register'));

    await (() => {
      expect(mockNavigate).toHaveBeenCalledWith('SignIn', {
        email: '',
      });
    });
  });

  test('should navigate to SignIn Screen with email after entering valid email address.', async () => {
    const mockedParams = {
      route: {params: {email: email}},
    };
    renderWithRedux(<RegistrationScreen />);

    const toClick = await screen.getByTestId('register');

    fireEvent(toClick, 'press');
    const email = await screen.getByTestId('emailID');

    const signInScreen = renderWithRedux(<SignInScreen {...mockedParams} />);

    expect(signInScreen).toBeTruthy();
  });
});
