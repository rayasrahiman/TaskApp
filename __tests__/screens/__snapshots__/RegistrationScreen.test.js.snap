// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationScreen should render Registration screen correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#CCCCCC",
        "borderBottomWidth": 2.5,
        "flexDirection": "row",
        "height": 200,
        "justifyContent": "space-between",
        "width": 750,
      }
    }
    testID="header"
  >
    <View
      style={
        Object {
          "marginLeft": 300,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../assets/Logo.png",
          }
        }
        style={
          Object {
            "height": 200,
            "width": 112.5,
          }
        }
      />
    </View>
  </View>
  <View
    testID="titleSubTitle"
  >
    <View
      style={
        Object {
          "marginHorizontal": 22.5,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#244379",
            "fontSize": 53.5,
            "fontWeight": "bold",
          }
        }
      >
        Welcome to the Pentair Home app!
      </Text>
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 22.5,
          "marginVertical": 26.5,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#838383",
            "fontSize": 26.5,
            "fontWeight": "bold",
          }
        }
      >
        Get the most out of your home's water. Enter your email to get started.
      </Text>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "marginHorizontal": 22.5,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000000",
            "fontSize": 26.5,
            "fontWeight": "bold",
          }
        }
      >
        Email Address
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#838383",
          "borderRadius": 13.5,
          "borderWidth": 2.5,
          "flexDirection": "row",
          "height": 80,
          "marginHorizontal": 22.5,
          "marginVertical": 10,
          "paddingHorizontal": 15,
        }
      }
    >
      <TextInput
        autoCorrect={false}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Email Address"
        style={
          Object {
            "color": "#000000",
            "flex": 1,
            "fontSize": 26.5,
          }
        }
        testID="emailID"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginHorizontal": 37.5,
        "marginTop": 400,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0E6E96",
          "borderRadius": 10,
          "height": 93.5,
          "justifyContent": "center",
          "opacity": 1,
          "width": "100%",
        }
      }
      testID="register"
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 40,
            "paddingVertical": 6.5,
          }
        }
      >
        Get Started
      </Text>
    </View>
  </View>
</View>
`;
